File: 2020/4/both.pl
#!/usr/bin/perl

use strict;
use integer;

my $fh;

die "* no input\n" if (!open($fh, '<', $ARGV[0] || 'input.txt'));

my $linenum = 0;
my @map;

my %fields =
(
	'byr' => [0x01, \&is_byr],
	'iyr' => [0x02, \&is_iyr],
	'eyr' => [0x04, \&is_eyr],
	'hgt' => [0x08, \&is_hgt],
	'hcl' => [0x10, \&is_hcl],
	'ecl' => [0x20, \&is_ecl],
	'pid' => [0x40, \&is_pid],
);


my $valid1 = 0;
my $present1 = 0;
my $valid2 = 0;
my $present2 = 0;

while (1)
{
	my $line = <$fh>;
	$linenum++;
	if ($line =~ /^\s*$/)
	{
		$valid1++ if ($present1 == 0x7f);
		$valid2++ if ($present2 == 0x7f);
		$present1 = 0;
		$present2 = 0;
		last if (!$line);
	}
	while ($line =~ /\b([a-z]{3}):([^ \r\n\t]+)/g)
	{
		my ($fld, $val) = ($1, $2);
		my $res = $fields{$fld};
		if ($res)
		{
			$present1 |= $res->[0];
			$present2 |= $res->[0] if ($res->[1]($val));
		}
	}
}

printf("Part 1 answer: %d\n", $valid1);
printf("Part 2 answer: %d\n", $valid2);

sub is_byr
{
	my ($x) = @_;
	if (($x =~ /^\d{4}$/) &&
	    ($x >= 1920) &&
	    ($x <= 2002))
	{
		return 1;
	}
	return 0;
}

sub is_iyr
{
	my ($x) = @_;
	return (($x =~ /^\d{4}$/) &&
		($x >= 2010) &&
		($x <= 2020));
}

sub is_eyr
{
	my ($x) = @_;
	return (($x =~ /^\d{4}$/) &&
		($x >= 2020) &&
		($x <= 2030));
}

sub is_hgt
{
	my ($x) = @_;
	if ($x =~ /^(\d+)(cm|in)$/)
	{
		return ($2 eq 'cm') ? (($1 >= 150) && ($1 <= 193)) :
				      (($1 >= 59) && ($1 <= 76));
	}
	return 0;
}

sub is_hcl
{
	my ($x) = @_;
	return ($x =~ /^#[0-9a-f]{6}$/);
}

sub is_ecl
{
	my ($x) = @_;
	return ($x =~ /^(amb|blu|brn|gry|grn|hzl|oth)$/);
}

sub is_pid
{
	my ($x) = @_;
	return ($x =~ /^\d{9}$/);
}

File: 2020/4/both.py
#!/usr/bin/python

import re
import string
import sys

def is_byr(x):
	if (x.isdigit()):
		x = int(x)
		if (x >= 1920) and (x <= 2002):
			return True
	return False

def is_iyr(x):
	if (x.isdigit()):
		x = int(x)
		if (x >= 2010) and (x <= 2020):
			return True
	return False

def is_eyr(x):
	if (x.isdigit()):
		x = int(x)
		if (x >= 2020) and (x <= 2030):
			return True
	return False

def is_hgt(x):
	if len(x) < 4:
		return False
	num = x[:-2]
	if num.isdigit():
		num = int(num)
		units = x[-2:]
		if units == 'cm':
			return ((num >= 150) and (num <= 193))
		if units == 'in':
			return ((num >= 59) and (num <= 76))
	return False

def is_hcl(x):
	if len(x) == 7 and x[0] == '#':
		return all(c in string.hexdigits for c in x[1:])
	return False

def is_ecl(x):
	return x in { 'amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth' }

def is_pid(x):
	return len(x) == 9 and x.isdigit()

infile = 'input.txt'
if (len(sys.argv) > 1):
	infile = sys.argv[1]

fh = open(infile, 'r');

fields = {
	'byr': (0x01, is_byr),
	'iyr': (0x02, is_iyr),
	'eyr': (0x04, is_eyr),
	'hgt': (0x08, is_hgt),
	'hcl': (0x10, is_hcl),
	'ecl': (0x20, is_ecl),
	'pid': (0x40, is_pid),
}

valid1 = 0
present1 = 0
valid2 = 0
present2 = 0
linenum = 0

r = re.compile(r"\b([a-z]{3}):([^ ]+)")

for line in fh:
	linenum += 1

	line = line.rstrip()
	if len(line) == 0:
		if present1 == 0x7f:
			valid1 += 1
		if present2 == 0x7f:
			valid2 += 1
		present1 = 0
		present2 = 0

	for m in r.finditer(line):
		if m.group(1) in fields:
			res = fields[m.group(1)]
			present1 |= res[0]
			if res[1](m.group(2)):
				present2 |= res[0]

if present1 == 0x7f:
	valid1 += 1
if present2 == 0x7f:
	valid2 += 1

print("Part 1 answer: %d" % valid1)
print("Part 2 answer: %d" % valid2)
